mixer/adapter/signalfx/signalfx.go

```diff
@@ -0,0 +1,188 @@
+var _ metric.HandlerBuilder = &builder{}
+var _ metric.Handler = &handler{}
+
+///////////////// Configuration-time Methods ///////////////
+
+// adapter.HandlerBuilder#Build
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+	confsByName := make(map[string]*config.Params_MetricConfig, len(b.config.Metrics))
+	for i := range b.config.Metrics {
+		confsByName[b.config.Metrics[i].Name] = b.config.Metrics[i]
+	}
+
+	env.Logger().Infof("Building SignalFx adapter")
+
+	h := &handler{
+		env:                env,
+		metricTypes:        b.metricTypes,
+		ingestURL:          b.config.IngestUrl,
+		accessToken:        b.config.AccessToken,
+		intervalSeconds:    b.config.IntervalSeconds,
+		metricConfigByName: confsByName,
+		logger:             env.Logger(),
+	}
+
+	return h, h.Init()
+}
+
+// adapter.HandlerBuilder#SetAdapterConfig
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.config = cfg.(*config.Params)
+
+	if b.config.IntervalSeconds == 0 {
+		b.config.IntervalSeconds = 10
+	}
+}
+
+// adapter.HandlerBuilder#Validate
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	if b.config.AccessToken == "" {
+		ce = ce.Appendf("access_token", "You must specify the SignalFx access token")
+	}
+	if len(b.config.Metrics) == 0 {
+		ce = ce.Appendf("metrics", "There must be at least one metric definition for this to be useful")
+	}
+
+	for i := range b.config.Metrics {
+		if b.config.Metrics[i].Type == config.NONE {
+			ce = ce.Appendf(fmt.Sprintf("metrics[%d].type", i), "type must be specified")
+		}
+
+		name := b.config.Metrics[i].Name
+		if len(name) == 0 {
+			ce = ce.Appendf(fmt.Sprintf("metrics[%d].name", i), "name must not be blank")
+			continue
+		}
+
+		if typ := b.metricTypes[name]; typ == nil {
+			ce = ce.Appendf(fmt.Sprintf("metrics[%d].name", i),
+				"Name %s does not correspond to a metric type registered in Istio", name)
+		}
+	}
+
+	return
+}
+
+// metric.HandlerBuilder#SetMetricTypes
+func (b *builder) SetMetricTypes(types map[string]*metric.Type) {
+	b.metricTypes = types
+}
+
+////////////////// Request-time Methods //////////////////////////
+
+func (h *handler) Init() error {
+	h.scheduler = sfxclient.NewScheduler()
+
+	h.scheduler.ReportingDelay(time.Duration(h.intervalSeconds) * time.Second)
+	h.scheduler.Sink.(*sfxclient.HTTPSink).AuthToken = h.accessToken
+
+	h.scheduler.ErrorHandler = func(err error) error {
+		return h.logger.Errorf("Error sending datapoints to SignalFx: %s", err.Error())
+	}
+
+	if h.ingestURL != "" {
+		h.scheduler.Sink.(*sfxclient.HTTPSink).DatapointEndpoint = fmt.Sprintf("%s/v2/datapoint", h.ingestURL)
+	}
+
+	h.ctx, h.cancel = context.WithCancel(context.Background())
+
+	h.registry = newRegistry(5 * time.Minute)
+	h.scheduler.AddCallback(h.registry)
+
+	h.env.ScheduleDaemon(func() {
+		err := h.scheduler.Schedule(h.ctx)
+		if err != nil {
+			h.logger.Infof("SignalFx scheduler shutdown with result %s", err.Error())
```

Comment:
Should this be using Errorf instead?

What's the state of the adapter is this returns with error != nil? Are things hosed at that point?

Reply:
Schedule returns an error whenever it is shutdown similar to `http.Server.ListenAndServe` so I didn't want to cause an error log on normal shutdown.  I'll add some logic to distinguish between normal shutdown and bad things happening.